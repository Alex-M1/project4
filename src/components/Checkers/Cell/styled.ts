import { IStyled } from 'src/components/_common_/types/styledTypes';
import { changeTheme } from 'src/helpers/changeTheme';
import styled from 'styled-components';

interface IStCellStyled extends IStyled {
  background: string;
}

export const StCell = styled.div <IStCellStyled>`
  width: 60px;
  height: 60px;
  border: 1px solid ${(props) => changeTheme(props, 'textColor')};
  margin: 1px;
  color: ${(props) => changeTheme(props, 'textColor')};
  background: ${(props) => changeTheme(props, 'gray')};
  transform: rotate(180deg);
  background: ${({ background }) => background};

  & div {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 40px;
    height: 40px;
    border-radius: 50%;
    cursor: grab;
    &::after {
      content: '';
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      width: 25px;
      height: 25px;
      border-radius: 50%;
    }
  }
`;
