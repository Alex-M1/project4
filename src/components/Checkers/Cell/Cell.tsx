import React from 'react';
import { useTheme } from 'src/components/hooks/useTheme';
import { VIEW_OPTIONS } from 'constants/constants';
import { ICheckerModel } from 'store/checkers/types';
import Checker from '../Checker';
import { StCell } from './styled';

export interface IProps {
  fieldItem: ICheckerModel,
  cellNumber: number,
  possibleCell?: boolean,
  doStep: (cellNum: number) => void,
  chooseCell: (cell: number) => void,
}

const Cell: React.FC<IProps> = ({
  doStep,
  fieldItem,
  cellNumber,
  chooseCell,
  possibleCell,
}) => {
  const { colors, theme } = useTheme();
  const cellBg = VIEW_OPTIONS.CELL_BACKGROUND;
  const cellHandler = () => chooseCell(cellNumber);
  const possibleHandler = () => doStep(cellNumber);
  const setBg = () => {
    if (possibleCell) return cellBg.POSSIBLE;
    return fieldItem.blackSquare ? cellBg.GRAY : cellBg.WHITE;
  };
  return (
    <StCell
      theme={theme}
      colors={colors}
      background={setBg()}
      onClick={possibleCell ? possibleHandler : cellHandler}
    >
      {
        fieldItem.checker
          ? <Checker isBlack={fieldItem?.checker.blackChecker} />
          : null
      }
    </StCell >
  );
};

export default Cell;
