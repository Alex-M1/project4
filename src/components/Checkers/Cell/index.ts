import { connect } from 'react-redux';
import { chooseCell, doStep } from 'store/checkers/actions';
import { getFieldItem, getPossibleStepsCells } from 'store/checkers/selectors';
import { AppStateType } from 'store/rootReducer';
import Cell, { IProps } from './Cell';

const mapStateToProps = (state: AppStateType, { cellNumber }: IProps) => ({
  fieldItem: getFieldItem(state, cellNumber),
  possibleCell: getPossibleStepsCells(state, cellNumber),
});

const mapDispatchToProps = {
  doStep,
  chooseCell,
};

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
