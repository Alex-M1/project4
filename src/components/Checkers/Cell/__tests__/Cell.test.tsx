import React from 'react';
import { shallow, mount } from 'enzyme';
import Cell from '../Cell';

describe('Cell', () => {
  describe('without Checker', () => {
    let props;
    beforeEach(() => {
      props = {
        fieldItem: {},
        cellNumber: 2,
        possibleCell: false,
        doStep: jest.fn(),
        chooseCell: jest.fn(),
      };
    });
    it('Should match snapshot', () => {
      const component = shallow(<Cell {...props} />);
      expect(component.html()).toMatchSnapshot();
    });
    it('should render StCell', () => {
      const component = mount(<Cell {...props} />);
      expect(component.find('styled__StCell')).toHaveLength(1);
    });
    it('should not render Checker', () => {
      const component = mount(<Cell {...props} />);
      expect(component.find('Checker')).not.toHaveLength(1);
    });
    it('should click StCell', () => {
      const component = mount(<Cell {...props} />);
      component.find('styled__StCell').simulate('click');
      expect(props.chooseCell).toHaveBeenCalled();
    });
  });
  describe('with Checker', () => {
    let props;
    beforeEach(() => {
      props = {
        fieldItem: { checker: true },
        cellNumber: 2,
        possibleCell: true,
        doStep: jest.fn(),
        chooseCell: jest.fn(),
      };
    });
    it('Should match snapshot', () => {
      const component = shallow(<Cell {...props} />);
      expect(component.html()).toMatchSnapshot();
    });
    it('should render StCell', () => {
      const component = mount(<Cell {...props} />);
      expect(component.find('styled__StCell')).toHaveLength(1);
    });
    it('should not render Checker', () => {
      const component = mount(<Cell {...props} />);
      expect(component.find('Checker')).toHaveLength(1);
    });
    it('should click StCell', () => {
      const component = mount(<Cell {...props} />);
      component.find('styled__StCell').simulate('click');
      expect(props.doStep).toHaveBeenCalled();
    });
  });
});
