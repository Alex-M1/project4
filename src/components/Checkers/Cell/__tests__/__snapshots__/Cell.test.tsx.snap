// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell white Cell Should match snapshot 1`] = `"<div class=\\"styled__StCell-gJrZHm fKDHEl\\"></div>"`;

exports[`Cell with Checker Should match snapshot 1`] = `"<div class=\\"styled__StCell-gJrZHm cngUHY\\"><div><div class=\\"styled__StCellWhite-eopeRB czkzjO\\"></div></div></div>"`;

exports[`Cell without Checker Should match snapshot 1`] = `"<div class=\\"styled__StCell-gJrZHm fKDHEl\\"></div>"`;
